include/ginkgo/core/preconditioner/isai.hpp

```diff
@@ -0,0 +1,206 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_PRECONDITIONER_ISAI_HPP_
+#define GKO_CORE_PRECONDITIONER_ISAI_HPP_
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Preconditioner namespace.
+ *
+ * @ingroup precond
+ */
+namespace preconditioner {
+
+
+/**
+ * This enum lists the types of the ISAI preconditioner.
+ *
+ * ISAI can either generate a lower triangular matrix, or an upper triangular
+ * matrix.
+ */
+enum struct isai_type { lower, upper };
+
+/**
+ * The Incomplete Sparse Approximate Inverse (ISAI) Preconditioner generates
+ * an approximate inverse matrix for a given lower triangular matrix L or upper
+ * triangular matrix U.
+ *
+ * Using the preconditioner computes $aiU * x$ or $aiL * x$ (depending on the
+ * type of the Isai) for a given vector x (may have multiple right hand sides).
+ * aiU and aiL are the approximate inverses for U and L respectively.
+ *
+ * The sparsity pattern used for the approximate inverse is the same as
+ * the sparsity pattern of the respective triangular matrix.
+ *
+ * For more details on the algorithm, see the paper
+ * <a href="https://doi.org/10.1016/j.parco.2017.10.003">
+ * Incomplete Sparse Approximate Inverses for Parallel Preconditioning</a>,
+ * which is the basis for this work.
+ *
+ * @tparam IsaiType  determines if the ISAI is generated for a lower triangular
+ *                   matrix or an upper triangular matrix
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup isai
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <isai_type IsaiType, typename ValueType, typename IndexType>
+class Isai : public EnableLinOp<Isai<IsaiType, ValueType, IndexType>> {
+    friend class EnableLinOp<Isai>;
+    friend class EnablePolymorphicObject<Isai, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using Csr = matrix::Csr<ValueType, IndexType>;
+    static constexpr isai_type type{IsaiType};
+
+    /**
+     * Returns the approximate inverse of the given matrix (either L or U,
+     * depending on the template parameter IsaiType).
+     *
+     * @returns the generated approximate inverse
+     */
+    std::shared_ptr<const Csr> get_system_matrix() const
```

Comment:
If this is the computed inverse, then I think the function should be named `get_matrix_inverse` or something like that. Or if this is actually returning the passed in `system_matrix`, then the documentation is incorrect

Reply:
Good point, it is returning the approximate inverse.
I will rename it to `get_approximate_inverse` to reduce confusion.